### ISSUE TITLE
"make-sized-byte-string: unsupported" error when calling `sha256` from module `sha`

### ISSUE THREAD

**Issue by khinsen** [racket/racket#2392]
Using today's snapshot of racket-cs (macOS version), I got compiler crashes when running the test suite of my program. They also occur when doing a simple `require` in the REPL. The output is attached below, the module being loaded is [available on GitHub](https://github.com/khinsen/leibniz/blob/pollen/leibniz/documents.rkt).

```
Welcome to Racket v7.1.0.2 [cs].
> (require leibniz/documents)
; make-sized-byte-string: unsupported [,bt for context]
> ,bt
; make-sized-byte-string: unsupported
;   context...:
;    raise-unsupported-error
;    /Users/racket/build/plt/build/cs/racket/collects/racket/contract/private/arrow-val-first.rkt:388:18
;    /Users/racket/build/plt/build/cs/racket/src/build/cs//compiled/expander.scm:42962:96: body of "/Users/hinsen/projects/leibniz/leibniz/documents.rkt"
;    /Users/racket/build/plt/build/cs/racket/src/build/cs//compiled/expander.scm:18817:13: for-loop
;    /Users/racket/build/plt/build/cs/racket/src/build/cs//compiled/expander.scm:19081:5: run-module-instance!125
;    /Users/racket/build/plt/build/cs/racket/src/build/cs//compiled/expander.scm:32657:5: perform-require!78
;    /Users/racket/build/plt/build/cs/racket/src/build/cs//compiled/expander.scm:30675:13: for-loop
;    /Users/racket/build/plt/build/cs/racket/src/build/cs//compiled/expander.scm:46353:5: body of top-level
;    call-with-values
;    /Users/racket/build/plt/build/cs/bundle/racket/share/pkgs/xrepl-lib/xrepl/xrepl.rkt:1477:0
;    /Users/racket/build/plt/build/cs/racket/collects/racket/repl.rkt:11:26
;    main.sps:614:4: thunk_6
```

**Comment by khinsen** [comment]
@samth That looks indeed like a reasonable explanation, but it's not my code that is calling `make-sized-byte-string`. The call is somewhere in the internals of Racket.

**Comment by khinsen** [comment]
Ahhh... Found it. The occurence of `expander.scm` fooled me into believing that the crash happened during macro expansion. But it's plain code execution at phase 0 inside my code that computes a checksum. Here's a minimal demonstration:

```
Welcome to Racket v7.1.0.2 [cs].
> (require sha)
> (sha256 #"abc")
; make-sized-byte-string: unsupported [,bt for context]
```

So the bug is in the `sha` module.

**Comment by samth** [comment]
The `sha` library needs to be fixed, but `racket/base` now provides `sha256-bytes` which does what you want.

**Comment by samth** [comment]
Also note that the `sha` library is by @greghendershott and is here: https://github.com/greghendershott/sha

**Comment by khinsen** [comment]
Oops,  I didn't realize that `sha` was not part of Racket. And yes, `sha256-bytes` replaces it perfectly. Thanks!

**Comment by greghendershott** [comment]
I'll fix this in the `sha` package. I will have it copy the bytes instead (but keeping using the faster `make-sized-byte-string` on older Rackets).

---

I'm not sure why RacketCS doesn't just do that -- have the new implementation of `make-sized-byte-string` copy instead?  (Yes, it's slower. Yes, programs will keep working. Yes, documentation can explain why it's slower, and people who notice/care can use another strategy if/when it matters.)

TL;DR: I don't understand why RacketCS throws an exception, as opposed to doing the correct thing more slowly?

**Comment by mflatt** [comment]
The `make-sized-byte-string` function is supposed to created a byte string that shares its content with the given pointer, so changes to the pointer are reflected in the bye string view and vice versa. RacketCS can't do that, because Chez Scheme's representation for byte strings doesn't have an indirection like traditional Racket's representation.

**Comment by greghendershott** [comment]
Ah, thank you for explaining.

Does [this commit](https://github.com/greghendershott/sha/commit/2496aef0fe1502b1855c4e818f110347a69d0203) seem like a reasonable compatibility strategy for me to use?

**Comment by samth** [comment]
You can also use `(system-type 'vm)` to tell the difference.

**Comment by greghendershott** [comment]
@samth Thanks for pointing that out. (Usually I prefer to check for a capability or behavior directly (e.g. using `dynamic-require` and/or `with-handlers`) rather than use a proxy like version numbers or labels. But in this case `system-type` is probably reasonable and reliable.)
